### PR TITLE
[WIP] Add Microsoft.EntityFrameworkCore.DocumentDb package

### LINKED ISSUE #8443: Cosmos DB provider
Note: This issue is for discussion and questions about the EF Core Cosmos provider.

Issue #12086 is tracking the work being done on the implementation.

----

Please consider creating entity framework core providers for Cosmos DB.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #8443

This is WIP. Early feedback is welcome.
Test time for functional tests on Emulator last time was ~5h.

SimpleQueryTests are passing.
It translates Predicate to SQL correctly. It supports TPH through discriminator.

### CODE REVIEW

File: build/dependencies.props [hunk 1]
```diff
@@ -5,6 +5,7 @@
   <PropertyGroup Label="Package Versions">
     <InternalAspNetCoreSdkPackageVersion>2.1.0-preview1-15626</InternalAspNetCoreSdkPackageVersion>
     <MicrosoftAspNetCoreHostingWebHostBuilderFactorySourcesPackageVersion>2.1.0-preview1-27807</MicrosoftAspNetCoreHostingWebHostBuilderFactorySourcesPackageVersion>
+    <MicrosoftAzureDocumentDBCorePackageVersion>1.7.1</MicrosoftAzureDocumentDBCorePackageVersion>
```
Comment (smitpatel) [inline build/dependencies.props]: @natemcmaster @ryanbrandenburg - This is new dependency to add. Where it should be added in Universe for automation?

File: samples/DocumentDbSample/Program.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+using Microsoft.Extensions.Logging;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.EntityFrameworkCore.Extensions;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+
+namespace DocumentDbSample
+{
+    public class Program
+    {
```
Comment (smitpatel) [inline samples/DocumentDbSample/Program.cs]: This is currently sample which makes it easier to expand (rather than finding the test for the scenario). It would be removed in future or can be converted to doc sample.

File: version.props [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ﻿<Project>
   <PropertyGroup>
     <VersionPrefix>2.1.0</VersionPrefix>
+    <ExperimentalProjectVersionPrefix>0.1.0</ExperimentalProjectVersionPrefix>
```
Comment (smitpatel) [inline version.props]: Current Version.

File: samples/DocumentDbSample/DocumentDbSample.csproj [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFrameworks>net461;netcoreapp2.0;netcoreapp2.1</TargetFrameworks>
+        <RuntimeFrameworkVersion>2.1.0-preview2-25624-02</RuntimeFrameworkVersion>
+        <NETStandardImplicitPackageVersion>2.1.0-preview2-25624-02</NETStandardImplicitPackageVersion>
```
Comment (natemcmaster) [inline samples/DocumentDbSample/DocumentDbSample.csproj]: You should remove these to properties. They are managed automatically via dependencies.props

File: build/dependencies.props (hunk 1 above)
Comment (natemcmaster) [reply build/dependencies.props]: https://github.com/aspnet/Universe/blob/dev/build/external-dependencies.props

**Comment by smitpatel** [comment]
@ajcvickers @divega - The syntax of query SQL generated would be following
```
@__predicateTuple_Item2_0='London'

SELECT c[""CustomerID""], c[""Address""], c[""City""], c[""CompanyName""], c[""ContactName""], c[""ContactTitle""], c[""Country""], c[""Fax""], c[""Phone""], c[""PostalCode""], c[""Region""]
FROM Customers c
WHERE c[""City""] = @__predicateTuple_Item2_0
```

While `c.CustomerID` is also valid but if the key name has space or `"` then it . syntax does not work and we have to generate `c["property name escaping double quote"]`
Should we do it always or only when it is needed?

Also in From clause, Customers is redudant. by default it always assumes its iterating over collection so just `From c` also works, or even using any other alias than c works. While `Customers c` is readable but if the collection name has space it is impossible to represent it in SQL. The query knows what to iterate on because query are sent to specific collection only. Should we avoid having name of collection altogether or do it only when needed?

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Update;
+
+namespace Microsoft.EntityFrameworkCore.Infrastructure
+{
+    public class DocumentDbClientService : IDocumentDbClientService
+    {
+        private static readonly string _remoteEndpoint = "https://localhost:8081";
+        private static readonly string _remoteKey
+            = "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==";
```
Comment (AndriySvyryd) [inline src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: Move to options

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbTransactionManager.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        {
+        }
+
+        public void ResetState()
+        {
+        }
+
+        public void RollbackTransaction()
+        {
+        }
```
Comment (AndriySvyryd) [inline src/EFCore.DocumentDb/Infrastructure/DocumentDbTransactionManager.cs]: Throw from all of these?

File: src/EFCore.DocumentDb/Properties/InternalsVisibleTo.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.EntityFrameworkCore.DocumentDb.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (AndriySvyryd) [inline src/EFCore.DocumentDb/Properties/InternalsVisibleTo.cs]: Why is this needed?

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: Moved to options coming from `UseDocumentDb` method. Was pending a push to the branch.

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbTransactionManager.cs (hunk 6 above)
Comment (smitpatel) [reply src/EFCore.DocumentDb/Infrastructure/DocumentDbTransactionManager.cs]: I copied that over from InMemory. I am not sure what should be behavior of these since I haven't added tests which uses this.

File: src/EFCore.DocumentDb/Properties/InternalsVisibleTo.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore.DocumentDb/Properties/InternalsVisibleTo.cs]: Copy over. Other projects have it. At present there are no unit tests.

File: EFCore.Runtime.sln [hunk 8]
```diff
@@ -3,18 +3,25 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
-	ProjectSection(SolutionItems) = preProject
-		test\Directory.Build.props = test\Directory.Build.props
-	EndProjectSection
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore", "src\EFCore\EFCore.csproj", "{715C38E9-B2F5-4DB2-8025-0C6492DEBDD4}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.Attributes", "src\EFCore.Attributes\EFCore.Attributes.csproj", "{92C1AEF8-D4FC-4631-98B1-F16FB7103398}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.Design", "src\EFCore.Design\EFCore.Design.csproj", "{D3D0A8E8-EC2F-4E01-8650-8554E186A66F}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EFCore.DocumentDb", "src\EFCore.DocumentDb\EFCore.DocumentDb.csproj", "{5E9511F2-945C-4DAE-B42C-C11DB048F0E8}"
```
Comment (anpete) [inline EFCore.Runtime.sln]: Noooooooooo :trollface:

File: samples/DocumentDbSample/Program.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
+using System;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.EntityFrameworkCore.Extensions;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+
+namespace DocumentDbSample
```
Comment (anpete) [inline samples/DocumentDbSample/Program.cs]: Remove "Sample" from here and directory?

File: src/EFCore.DocumentDb/Extensions/DocumentDbLoggerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            builder
+                .Append(parameter.Name)
+                .Append("=");
+
+            FormatParameterValue(builder, parameter.Value);
+
+            return builder.ToString();
+        }
+
+        private static void FormatParameterValue(StringBuilder builder, object parameterValue)
```
Comment (anpete) [inline src/EFCore.DocumentDb/Extensions/DocumentDbLoggerExtensions.cs]: Core extension method?

File: src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                Id = GetId(entry)
+            };
+            foreach (var property in _entityType.GetProperties())
+            {
+                document.SetPropertyValue(property.Name, entry.GetCurrentValue(property));
+            }
+
+            try
+            {
+                await _documentClient.CreateDocumentAsync(
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs]: Tail await call. Remove try and return Task directly. Or is the try needed for some reason?

File: src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                throw e;
+            }
+        }
+
+        public async Task DeleteAsync(IUpdateEntry entry)
+        {
+            try
+            {
+                await _documentClient.DeleteDocumentAsync(
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs]: Tail await call. Remove try and return Task directly. Or is the try needed for some reason?

File: src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+                Id = GetId(entry)
+            };
+            foreach (var property in _entityType.GetProperties())
+            {
+                document.SetPropertyValue(property.Name, entry.GetCurrentValue(property));
+            }
+
+            try
+            {
+                await _documentClient.ReplaceDocumentAsync(
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentCollectionService.cs]: Tail await call. Remove try and return Task directly. Or is the try needed for some reason?

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        private readonly IDiagnosticsLogger<DbLoggerCategory.Database.Command> _logger;
+        private readonly IDocumentCollectionServiceFactory _documentCollectionServiceFactory;
+        private Dictionary<IEntityType, IDocumentCollectionService> _collections
+            = new Dictionary<IEntityType, IDocumentCollectionService>();
+
+        public DocumentDbClientService(IDiagnosticsLogger<DbLoggerCategory.Database.Command> logger,
+            IDbContextOptions dbContextOptions,
+            IDocumentCollectionServiceFactory documentCollectionServiceFactory)
+        {
+            var option = dbContextOptions.Extensions.OfType<DocumentDbOptionsExtension>().Single();
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: nit: options

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+            _databaseId = option.DatabaseName;
+            _documentClient = new DocumentClient(option.ServiceEndPoint, option.AuthKeyOrResourceToken);
+            _logger = logger;
+            _documentCollectionServiceFactory = documentCollectionServiceFactory;
+        }
+
+        public DocumentClient Client => _documentClient;
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: TODO: Add API check tests at some point

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                {
+                    _collections.Add(
+                        entityType,
+                        documentCollectionService = _documentCollectionServiceFactory.Create(this, entityType));
+                }
+
+                switch (entry.EntityState)
+                {
+                    case EntityState.Added:
+                        await documentCollectionService.CreateAsync(entry);
```
Comment (anpete) [inline src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: Consider changing IDocumentCollectionService to have a single Update method that has this switch in its impl.

File: src/EFCore.DocumentDb/Metadata/DocumentDbAnnotationsBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+            ConfigurationSource = configurationSource;
+        }
+
+        public virtual ConfigurationSource ConfigurationSource { get; }
+        public virtual InternalMetadataBuilder MetadataBuilder { get; }
+        public override bool SetAnnotation(
+            string relationalAnnotationName,
+            object value)
+            => MetadataBuilder.HasAnnotation(relationalAnnotationName, value, ConfigurationSource);
+        public override bool CanSetAnnotation(
```
Comment (anpete) [inline src/EFCore.DocumentDb/Metadata/DocumentDbAnnotationsBuilder.cs]: nit: blank like before methods.

File: test/EFCore.DocumentDb.FunctionalTests/Query/SimpleQueryDocumentDbTest.cs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+            NorthwindQueryDocumentDbFixture<NoopModelCustomizer> fixture,
+            // ReSharper disable once UnusedParameter.Local
+            ITestOutputHelper testOutputHelper)
+            : base(fixture)
+        {
+            ClearLog();
+            Fixture.TestSqlLoggerFactory.SetTestOutputHelper(testOutputHelper);
+        }
+
+        public override void Select_Property_when_non_shadow()
```
Comment (anpete) [inline test/EFCore.DocumentDb.FunctionalTests/Query/SimpleQueryDocumentDbTest.cs]: Need any AssertSql?

File: src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs (hunk 14 above)
Comment (smitpatel) [reply src/EFCore.DocumentDb/Infrastructure/DocumentDbClientService.cs]: but its single! :trollface:

File: src/EFCore.DocumentDb/Metadata/DocumentDbAnnotationsBuilder.cs (hunk 17 above)
Comment (smitpatel) [reply src/EFCore.DocumentDb/Metadata/DocumentDbAnnotationsBuilder.cs]: Formatting of "light-weight annotations framework" is pending. Will be doing once some annotations are in.

**Comment by smitpatel** [comment]
Superseded by #12619 which adds almost all features which were prototyped in this PR.
